### PR TITLE
add requiredModules in widget model and inject them when generating page

### PR DESCRIPTION
**Pull request by cpuy**
Widgets declare angular modules they need.
Then hey are then injected in main app module dependencies while generating page

### CODE REVIEW

**Comment by cpuy** [comment]
See https://github.com/bonitasoft/bonita-form-builder/pull/957

File: backend/webapp/src/main/java/org/bonitasoft/web/designer/model/widget/Widget.java [hunk 1]
```diff
@@ -55,6 +56,7 @@
     @Valid
     private List<Property> properties = new ArrayList<>();
     private Map<String, List<Identifiable>> usedBy; // list of element that use this widget
+    private Set<String> requiredModules;
```
Comment (javamind) [inline backend/webapp/src/main/java/org/bonitasoft/web/designer/model/widget/Widget.java]: Why you don't use a List ?

File: backend/webapp/src/main/java/org/bonitasoft/web/designer/rendering/TemplateEngine.java [hunk 2]
```diff
@@ -31,6 +32,7 @@
     public TemplateEngine(String template) {
         handlebars = new Handlebars(new ClassPathTemplateLoader("/", ""));
         handlebars.registerHelper("json", Jackson2Helper.INSTANCE);
+        handlebars.registerHelper("join", StringHelpers.join);
```
Comment (javamind) [inline backend/webapp/src/main/java/org/bonitasoft/web/designer/rendering/TemplateEngine.java]: For my knowledge, can you explain me the utility ?

File: backend/webapp/src/main/java/org/bonitasoft/web/designer/model/widget/Widget.java (hunk 1 above)
Comment (javamind) [inline backend/webapp/src/main/java/org/bonitasoft/web/designer/model/widget/Widget.java]: Ok I understand. Several widget can add the same module..

File: backend/webapp/src/main/java/org/bonitasoft/web/designer/rendering/TemplateEngine.java (hunk 2 above)
Comment (javamind) [inline backend/webapp/src/main/java/org/bonitasoft/web/designer/rendering/TemplateEngine.java]: I see below in the PR the utility...

**Comment by javamind** [comment]
Cool.. Very nice for ui-designer extensibility
This add is only for our widgets or in a next PR we will add the feature for the custom widgets ?

File: backend/webapp/src/main/java/org/bonitasoft/web/designer/model/widget/Widget.java (hunk 1 above)
Comment (cpuy) [inline backend/webapp/src/main/java/org/bonitasoft/web/designer/model/widget/Widget.java]: :+1:

File: backend/webapp/src/main/java/org/bonitasoft/web/designer/rendering/TemplateEngine.java (hunk 2 above)
Comment (cpuy) [inline backend/webapp/src/main/java/org/bonitasoft/web/designer/rendering/TemplateEngine.java]: Used in .hbs templates

**Comment by cpuy** [comment]
For now this is just for our widgets, but we have just to add a new field in custom widget editor and that's it.
This is a frequently asked feature which will be useful with widgets assets

**Comment by javamind** [comment]
Ok the merge is coming...

**Pull request merged**